### PR TITLE
Store span names & types, input names & types as internal trace tag

### PR DESCRIPTION
**Pull request by jessechancy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/jessechancy/mlflow/pull/12015?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12015/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12015
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

When a trace is logged from the MLflow client to Databricks, we should set amlflow.traceSpans tag (see https://github.com/databricks/universe/pull/559943 )  via the SetTraceTag API call that includes the following information for each span:

The span name

The span type

The span input names (only if the span outputs are a dict)

The span output names (only if the span inputs are a dict)

We should store this as JSON with keys name (value is string), type (value is string), inputs (value is list of string), outputs (value is list of string). To save space, we should not pretty print the JSON - it should be as compact as possible.

If the JSON is too large, the backend will throw an INVALID_PARAMETER_EXCEPTION (see https://github.com/databricks/universe/pull/559943 ). We should try / catch the tag logging; if this exception is encountered, we should just skip logging this tag (the trace should still be logged, the user shouldn’t see any exceptions). It’s important to use SetTraceTag, not EndTrace, to set the tag because we don’t want EndTrace to fail if the tag length is too long. We should not hardcode the tag length check in the client, since the maximum tag length may differ based on the backend.

This tag will be used by the UI to populate a dropdown of span fields (inputs, outputs) that users can select, allowing them to extract the fields and view them in a table.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/tracing/trace_manager.py [hunk 1]
```diff
@@ -139,6 +139,15 @@ def get_request_id_from_trace_id(self, trace_id: int) -> Optional[str]:
         """
         return self._trace_id_to_request_id.get(trace_id)
 
+    def get_mlflow_trace_from_trace(self, request_id: int) -> Optional[Trace]:
```
Comment (serena-ruan) [inline mlflow/tracing/trace_manager.py:142]: Can we rename this to get_mlflow_trace? from_trace sounds a bit weird

File: mlflow/tracking/client.py [hunk 2]
```diff
@@ -648,6 +649,29 @@ def end_trace(
...
+        # https://databricks.atlassian.net/browse/ML-40306
+        parsed_spans = []
+        for span in trace_data.spans:
+            parsed_span = {}
+
+            parsed_span["name"] = span.name
+            parsed_span["type"] = span.get_attribute(SpanAttributeKey.SPAN_TYPE)
+            span_inputs = span.get_attribute(SpanAttributeKey.INPUTS)
+            if span_inputs and isinstance(span_inputs, dict):
+                parsed_span["inputs"] = list(span_inputs.keys())
```
Comment (serena-ruan) [inline mlflow/tracking/client.py:663]: What if span_inputs is not a dict, do we want to store it or not?

Comment (serena-ruan) [reply mlflow/tracking/client.py:663]: Same for outputs

File: mlflow/tracing/export/mlflow.py [hunk 3]
```diff
@@ -80,3 +80,8 @@ def _log_trace(self, trace: Trace):
             self._client._upload_ended_trace_info(trace.info)
         except Exception as e:
             _logger.debug(f"Failed to log trace to MLflow backend: {e}", exc_info=True)
+
+        try:
+            self._client._upload_trace_spans_as_tag(trace.info, trace.data)
```
Comment (serena-ruan) [inline mlflow/tracing/export/mlflow.py:85]: Can we put this above end_trace? It seems more reasonable that we update tags before ending trace (which includes status).

File: mlflow/tracking/client.py (hunk 2 above)
Comment (jessechancy) [reply mlflow/tracking/client.py:663]: If span inputs or outputs is not a dict, we don't want to save it

File: mlflow/tracing/export/mlflow.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/tracing/export/mlflow.py:85]: I think that's fair

**Review by dbczumar — APPROVED** [review]
LGTM once comments from @serena-ruan are addressed! Thanks @jessechancy !

**Pull request merged**